packages/cmf/src/sagaRouter/router.js

```diff
@@ -30,25 +30,47 @@ function shouldRestartSaga(maybeSaga, match) {
 	return false;
 }
 
+function parseSagaState(routeFragments, sagas, currentLocation, index) {
+	const routeFragment = routeFragments[index];
+	return {
+		routeFragment,
+		match: matchPath(currentLocation.pathname, { path: routeFragment }),
+		maybeSaga: sagas[routeFragment],
+	};
+}
+
 export default function* routerSaga(history, routes) {
 	const sagas = {};
 	while (true) {
 		yield take('@@router/LOCATION_CHANGE');
 		const currentLocation = history.getCurrentLocation();
 		const routeFragments = Object.keys(routes);
 		for (let index = 0; index < routeFragments.length;) {
-			const routeFragment = routeFragments[index];
-			const match = matchPath(currentLocation.pathname, { path: routeFragment });
-			const maybeSaga = sagas[routeFragment];
-			if (shouldStartSaga(maybeSaga, match)) {
+			const { routeFragment, match, maybeSaga } = parseSagaState(
+				routeFragments,
+				sagas,
+				currentLocation,
+				index,
+			);
+			if (shouldCancelSaga(maybeSaga, match)) {
+				yield cancel(maybeSaga.saga);
+			} else if (shouldRestartSaga(maybeSaga, match)) {
+				yield cancel(maybeSaga.saga);
 				sagas[routeFragment] = {
 					saga: yield spawn(routes[routeFragment], match.params),
 					match,
 				};
-			} else if (shouldCancelSaga(maybeSaga, match)) {
-				yield cancel(maybeSaga.saga);
-			} else if (shouldRestartSaga(maybeSaga, match)) {
-				yield cancel(maybeSaga.saga);
+			}
+			index += 1;
+		}
+		for (let index = 0; index < routeFragments.length;) {
```

Comment:
Instead of running the loop twice, can you populate a sagaQueue in the first loop where you store the saga to start (routes/params/match) ?
Then after the loop, you only loop on this sagaQueue to start the sagas, avoiding `parseSagaState` computation on each element twice.